File: internal/provider/capability.go
// Package provider 定义模型能力、渠道适配器与调用传输层。
package provider

import "strings"

// Capability 是统一能力枚举，避免散落的字符串判断（docs/design/03 §2.3）。
type Capability string

// 能力枚举。
const (
	CapImageGenerate Capability = "image.generate"
	CapImageEdit     Capability = "image.edit"
	CapImageUpscale  Capability = "image.upscale"
	CapVideoGenerate Capability = "video.generate"
	CapTextGenerate  Capability = "text.generate"
	CapTextTools     Capability = "text.tools"
	CapAudioGenerate Capability = "audio.generate"
	CapModelList     Capability = "model.list"
)

// All 返回全部能力（供前端与文档校验）。
func All() []Capability {
	return []Capability{CapImageGenerate, CapImageEdit, CapImageUpscale, CapVideoGenerate,
		CapTextGenerate, CapTextTools, CapAudioGenerate, CapModelList}
}

// Valid 校验能力名。
func Valid(c Capability) bool {
	for _, x := range All() {
		if x == c {
			return true
		}
	}
	return false
}

// Parse 把字符串解析为能力（未知返回 false）。
func Parse(s string) (Capability, bool) {
	c := Capability(strings.TrimSpace(s))
	return c, Valid(c)
}

// ResourceKind 把能力映射到资源类型（用于画布端口类型校验）。
func (c Capability) ResourceKind() string {
	switch c {
	case CapImageGenerate, CapImageEdit, CapImageUpscale:
		return "image"
	case CapVideoGenerate:
		return "video"
	case CapTextGenerate, CapTextTools:
		return "text"
	case CapAudioGenerate:
		return "audio"
	}
	return "file"
}

File: web/src/features/settings/capability.ts
/**
 * 按模型名关键词猜测能力（对齐 docs/design/10 §8.3）。
 *
 * 与后端的 `GuessCapabilities` 保持同一套关键词——两边判定不一致时，
 * 会出现「前端显示是生图模型、后端不给它生图凭据」这种混乱。
 * 因此这里的关键词表与 `internal/provider/service.go` 的
 * `GuessCapabilities` 一一对应，并由 i18n 一致性测试同类的思路维护。
 *
 * 猜测只是默认值：用户可以覆盖。猜错不该导致「不能选」，只该导致「默认勾选错」。
 */
export function guessCapabilities(modelId: string): string[] {
  const m = (modelId ?? "").toLowerCase();
  if (m.includes("upscale") || m.includes("superres")) return ["image.upscale"];
  if (m.includes("embed") || m.includes("moderation")) return ["text.tools"];
  if (m.includes("tts") || m.includes("audio") || m.includes("speech"))
    return ["audio.generate"];
  if (
    m.includes("video") ||
    m.includes("veo") ||
    m.includes("sora") ||
    m.includes("kling")
  ) {
    return ["video.generate"];
  }
  if (
    m.includes("dall") ||
    m.includes("image") ||
    m.includes("flux") ||
    m.includes("sd") ||
    m.includes("imagen") ||
    m.includes("gpt-image")
  ) {
    return ["image.generate", "image.edit"];
  }
  return ["text.generate", "model.list"];
}
